### DISCUSSION TITLE
How to check if a username is already in use when a user is updating their profile

### DISCUSSION THREAD

**Discussion by shyakadavis** [pocketbase/pocketbase#4157] [Q&A] [answered]
Hi;

Say I want a user to be able to update their `name` and `username` on their profiles, how would I first check that the new `username` is not already in use?

Took me a while to get here, but I just learned about `locked` rules (handy feature) whereby I have to bypass it to first access the `users` collection.

From this, https://github.com/pocketbase/pocketbase/discussions/1018#discussioncomment-4128437, I was very curious/worried if not doing that optional bypass won't override the current user session with admin privileges, hence a security disaster?

If, however, this is the recommended way, how do I immediately log out that admin after doing my check?

I was thinking of adding 2 new .env vars, `PUBLIC_ADMIN_EMAIL` & `PUBLIC_ADMIN_PASSWORD` but I'm not sure about the best course of action here.

For context, I am using SvelteKit and doing those checks in form actions.

Thank you for the help.

**Comment by shyakadavis** [comment]
This is my `+page.server.ts` if it helps:

```ts
import { fail, redirect, type NumericRange, error } from '@sveltejs/kit';
import { message, setError, superValidate } from 'sveltekit-superforms/server';
import { ClientResponseError } from 'pocketbase';

import { profile_schema, type ProfileData } from '$lib/types/user';

export const load = async ({ locals }) => {
	if (!locals.user) redirect(302, '/auth/login');

	const profile = (await locals.pb.collection('users').getOne(locals.user.id)) as ProfileData;

	if (!profile) error(404);

	const form = await superValidate(profile, profile_schema);

	return { form, profile };
};

export const actions = {
	default: async (event) => {
		if (!event.locals.user) redirect(302, '/auth/login');

		const form = await superValidate(event, profile_schema);

		if (!form.valid) return fail(400, { form });

		try {
			const filter = `username="${form.data.username}"`;

			const user = await event.locals.pb.collection('users').getFirstListItem(filter);

			if (user && user.id !== event.locals.user.id) {
				return setError(form, 'username', 'Username already in use.');
			}

			await event.locals.pb.collection('users').update(event.locals.user.id, form.data);
		} catch (error) {
			if (error instanceof ClientResponseError) {
				console.error({ status: error.status, message: error.message });
				console.table({ ...error.response });
				return message(form, error.message, { status: error.status as NumericRange<400, 599> });
			}
			console.error(error);
			return message(form, 'An unknown error occurred', { status: 500 });
		}

		return message(form, 'Profile updated successfully');
	}
};
```

**Comment by ganigeorgiev** [comment, marked as answer]
> How to check if a username is already in use when a user is updating their profile

You don't have to manually check for duplicated usernames. The `username` field has builtin unique validator that should return an error on update if you are trying to set a value that already exists (_in the future there will be also policies to prevent usernames reuse and to forbid using a deleted/old username that has recently changed_).

> I was thinking of adding 2 new .env vars, PUBLIC_ADMIN_EMAIL & PUBLIC_ADMIN_PASSWORD but I'm not sure about the best course of action here.

I'm not sure that I understand what is your goal, but note that admins have access and can modify everything (including your db schema) and it is safe to be authenticated as admin **only** in a server-side context (in this case it is also safe to have a single global SDK instance; one such example could be https://github.com/pocketbase/pocketbase/discussions/3792#discussioncomment-7621130).

**Reply by shyakadavis** [reply]
> The username field has builtin unique validator that should return an error on update

Ah, got it.

Then I'm good with this, thank you.

```ts
catch (error) {
			if (error instanceof ClientResponseError) {
				if (error.response.data.username.code === 'validation_invalid_username') {
					return setError(form, 'username', error.response.data.username.message);
				}
				return message(form, error.message, { status: error.status as NumericRange<400, 599> });
			}
```

> I'm not sure that I understand what is your goal

Never mind that; I was just worried of doing another auth `await pb.admins.authWithPassword(...)` when I already have  another user logged in, but I'm glad to know that:
> it is safe to be authenticated as admin only in a server-side context

I don't 100% get it, though, but I'll read more of that thread, and if I ever run into an issue with this, I'll be sure to ask for help.

Thank you.
